Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,442 @@
+        }
+        
+        /// <summary>
+        /// Confines input 2D point within the confined area.
+        /// </summary>
+        /// <param name="positionToConfine">2D point to confine</param>
+        /// <returns>Confined position</returns>
+        private Vector2 ConfinePoint(Vector2 positionToConfine, in List<List<Vector2>> pathCache,
+            in Transform localToWorld, in Vector2 offset)
+        {
+            if (ShrinkablePolygon.IsInside(pathCache, positionToConfine, localToWorld, offset))
+            {
+                return Vector2.zero;
+            }
+
+            Vector2 closest = positionToConfine;
+            float minDistance = float.MaxValue;
+            for (int i = 0; i < pathCache.Count; ++i)
+            {
+                int numPoints = pathCache[i].Count;
+                if (numPoints > 0)
+                {
+                    Vector2 v0 = localToWorld.TransformPoint(pathCache[i][numPoints - 1] + offset);
+                    for (int j = 0; j < numPoints; ++j)
+                    {
+                        Vector2 v = localToWorld.transform.TransformPoint(pathCache[i][j] + offset);
+                        Vector2 c = Vector2.Lerp(v0, v, positionToConfine.ClosestPointOnSegment(v0, v));
+                        float distance = Vector2.SqrMagnitude(positionToConfine - c);
+                        if (distance < minDistance)
+                        {
+                            minDistance = distance;
+                            closest = c;
+                        }
+                        v0 = v;
+                    }
+                }
+            }
+            return closest - positionToConfine;
+        }
+        
+        internal static readonly float m_bakedConfinerResolution = 0.005f;
+        private class VcamExtraState
+        {
+            public Vector3 m_previousDisplacement;
+            public Vector3 m_dampedDisplacement;
+            public VcamShapeCache m_vcamShapeCache;
+            
+            /// <summary> Contains all the cache items that are dependent on something in the vcam. </summary>
+            internal struct VcamShapeCache
+            {
+                public List<List<Vector2>> m_path;
+                
+                private float m_frustumHeight;
+                private bool m_orthographic;
+
+                /// <summary>
+                /// Check that the path cache was converted from the current confiner cache, or
+                /// converts it if the frustum height was changed.
+                /// </summary>
+                /// <param name="confinerStateChanged">Confiner cache was changed</param>
+                /// <param name="frustumHeight">Camera frustum height</param>
+                public void ValidateCache(
+                    in ConfinerOven confinerBaker, in bool confinerStateChanged, 
+                    in float frustumHeight, in bool orthographic, 
+                    VcamExtraState extra)
```

Comment:
wtf?  what's wrong with "this" ?

Reply:
wtf indeed! :D it looks like I am tired. Fixed!